### ISSUE TITLE
`libc++.modules.json` resource does not exist

### ISSUE THREAD

**Issue by h-2** [llvm/llvm-project#120215]
I am trying this C++ Modules PoC:

CMakeLists.txt:
```CMake
cmake_minimum_required(VERSION 3.30)

set(CMAKE_CXX_STANDARD 23)
set(CMAKE_CXX_STANDARD_REQUIRED YES)

set(CMAKE_EXPERIMENTAL_CXX_IMPORT_STD "0e5b6991-d74f-4b3d-a41c-cf096e0b2508")
set(CMAKE_CXX_MODULE_STD ON)

project(example LANGUAGES CXX)
add_executable(main main.cpp)
```

main.cpp:
```cpp
import std;

int main()
{
        std::print("foobar\n");
}
```

I have the following relevant packages installed:
```
ii  clang-19                                                 1:19.1.5~++20241203083347+ab4b5a2db582-1~exp1~20241203083401.69 amd64        C, C++ and Objective-C compiler
ii  cmake                                                    3.30.2-0kitware1ubuntu24.04.1                                   amd64        cross-platform, open-source make system
ii  cmake-data                                               3.30.2-0kitware1ubuntu24.04.1                                   all          CMake data files (modules, templates and documentation
ii  libc++-19-dev:amd64                                      1:19.1.5~++20241203083347+ab4b5a2db582-1~exp1~20241203083401.69 amd64        LLVM C++ Standard library (development files)
ii  libc++1-19:amd64                                         1:19.1.5~++20241203083347+ab4b5a2db582-1~exp1~20241203083401.69 amd64        LLVM C++ Standard library
ii  libc++abi-19-dev:amd64                                   1:19.1.5~++20241203083347+ab4b5a2db582-1~exp1~20241203083401.69 amd64        LLVM low level support for a standard C++ library (development files)
ii  libc++abi1-19:amd64
```

I build the project with:
`cmake ../ -DCMAKE_CXX_COMPILER=clang++-19 -DCMAKE_CXX_FLAGS="-stdlib=libc++" -DCMAKE_BUILD_TYPE=Debug -G Ninja`

It outputs:
```-- The CXX compiler identification is Clang 19.1.5
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++-19 - skipped
-- Detecting CXX compile features
CMake Warning (dev) at /usr/share/cmake-3.30/Modules/Compiler/CMakeCommonCompilerMacros.cmake:248 (cmake_language):
  CMake's support for `import std;` in C++23 and newer is experimental.  It
  is meant only for experimentation and feedback to CMake developers.
Call Stack (most recent call first):
  /usr/share/cmake-3.30/Modules/CMakeDetermineCompilerSupport.cmake:113 (cmake_create_cxx_import_std)
  /usr/share/cmake-3.30/Modules/CMakeTestCXXCompiler.cmake:83 (CMAKE_DETERMINE_COMPILER_SUPPORT)
  CMakeLists.txt:10 (project)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Detecting CXX compile features - done
-- Configuring done (0.4s)
CMake Error in CMakeLists.txt:
  The "CXX_MODULE_STD" property on the target "main" requires that the
  "__CMAKE::CXX23" target exist, but it was not provided by the toolchain.
  Reason:

    `libc++.modules.json` resource does not exist


-- Generating done (0.0s)
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

The requested file **is** installed by the official apt-package; it is located at: `/usr/lib/llvm-19/lib/libc++.modules.json`

But clang doesn't seem to be able to find it there.

**Comment by tsurumi-yizhou** [comment]
The same to me.

**Comment by ChuanqiXu9** [comment]
What's your environment? If it is mac m1, it, this is duplicated with https://github.com/llvm/llvm-project/issues/109895

**Comment by h-2** [comment]
> What's your environment?

```
Distributor ID: Ubuntu
Description:    Ubuntu 24.04.1 LTS
Release:        24.04
Codename:       noble
```

**Comment by ChuanqiXu9** [comment]
For such environment sensitive problem, I may not be able to fix it actually.

@mathstuf I am wondering if CMake can offer an option as `STD_MODULE_JSON_PATH` so that users can workaround when the compiler has problems in different environments.

**Comment by ChuanqiXu9** [comment]
If anyone on Ubuntu is willing to look into this, I'll be appreciated. This should majorly be a driver problem with different path pattern in different platforms.

**Comment by tsunho12** [comment]
This is because `clang++ -print-file-name=libc++.modules.json` does not return the full path of “libc++.modules.json” (I don’t know why).

But the workaround is here. On Intel/apple silicon Mac, seem Linux haven't Clang-CXX-CXXImportStd.cmake, this file just for Mac, and other file for Windows, Linux now have't like this configuration file.
– Open the file “some path/cmake/some version/share/cmake/Modules/Compiler/Clang-CXX-CXXImportStd.cmake”
– Change line 13
– ” -print-file-name=libc++.modules.json” to ” -print-file-name=../../c++/libc++.modules.json”

**Comment by mathstuf** [comment]
> But the workaround is here. On Intel/apple silicon Mac, seem Linux haven't Clang-CXX-CXXImportStd.cmake, this file just for Mac, and other file for Windows, Linux now have't like this configuration file.

It affects Linux too. The core issue here is that the installation layout and the runtime expectations for the file are mismatched on macOS. CMake's code is fine, but it can host such a workaround in the meantime.

**Comment by h-2** [comment]
Yes, as I pointed out, this issue is not about macOS, it is about the official LLVM deb packages from apt.llvm.org on the most recent LTS version of Ubuntu.

**Comment by mpusz** [comment]
I have the same problem with clang-19 apt package on my Ubuntu :-(

**Comment by mordante** [comment]
Which version of libc++ do you have installed?
Can you post the path where `libc++.modules.json` is installed on your system?

**Comment by mpusz** [comment]
@mordante, I just played with this a bit more, and now I can't repro. I think that my issue was when I had libc++-18 installed but tried to compile the code with modules with clang-19. After I installed libc++-19 I did not have this error, but I couldn't compile my code because of the [compiler bugs](https://github.com/llvm/llvm-project/issues/110231). So I tried, clang-20 with libc++20, but this has yet another bug on the mainline. After that, I reverted back to clang-19 and libc++-18 and I can't repro anymore.

It is a pity that we can't install several libc++ versions in parallel, so I can't easily test various clang versions in one process without the need to reinstall libc++ libraries all the time.

**Comment by mordante** [comment]
> It is a pity that we can't install several libc++ versions in parallel, so I can't easily test various clang versions in one process without the need to reinstall libc++ libraries all the time.

@mpusz I don't know how hard that is to do for the Debian/Ubuntu packages. You could consider filing an issue with this request. Another solution might be using a Docker images with the different versions of clang/libc++.

**Comment by ChuanqiXu9** [comment]
@mordante hi, would you like to take a look?

**Comment by bpavuk** [comment]
having same issue on NixOS, but it's `libstdc++.modules.json` that is missing for me, as I link against that.
